rabbit/src/lib.rs

```diff
@@ -0,0 +1,515 @@
+    let s47_32 = ((state.state_vars[2]) ^ (state.state_vars[7] >> 16)) as u16;
+    let s63_48 = ((state.state_vars[2] >> 16) ^ (state.state_vars[5])) as u16;
+    let s79_64 = ((state.state_vars[4]) ^ (state.state_vars[1] >> 16)) as u16;
+    let s95_80 = ((state.state_vars[4] >> 16) ^ (state.state_vars[7])) as u16;
+    let s111_96 = ((state.state_vars[6]) ^ (state.state_vars[3] >> 16)) as u16;
+    let s127_112 = ((state.state_vars[6] >> 16) ^ (state.state_vars[1])) as u16;
+
+    s[0x0] = s15_0 as u8;
+    s[0x1] = (s15_0 >> 8) as u8;
+    s[0x2] = s31_16 as u8;
+    s[0x3] = (s31_16 >> 8) as u8;
+    s[0x4] = s47_32 as u8;
+    s[0x5] = (s47_32 >> 8) as u8;
+    s[0x6] = s63_48 as u8;
+    s[0x7] = (s63_48 >> 8) as u8;
+    s[0x8] = s79_64 as u8;
+    s[0x9] = (s79_64 >> 8) as u8;
+    s[0xA] = s95_80 as u8;
+    s[0xB] = (s95_80 >> 8) as u8;
+    s[0xC] = s111_96 as u8;
+    s[0xD] = (s111_96 >> 8) as u8;
+    s[0xE] = s127_112 as u8;
+    s[0xF] = (s127_112 >> 8) as u8;
+
+    return s;
+}
+
+/// Rabbit stream cipher state.
+pub struct Rabbit {
+    master_state: State,
+    state: State,
+    buf: [u8; 16],
+    buf_idx: u8,
+    buf_num: u64,
+}
+
+impl Rabbit {
+    /// Setupы given `key` on an empty rabbit state.
+    pub fn new(key: [u8; KEY_BYTE_LEN]) -> Rabbit {
+        let mut state = Default::default();
+        setup_key(&mut state, key);
+        Rabbit {
+            master_state: state.clone(),
+            state,
+            buf: [0; 16],
+            buf_idx: 0x10,
+            buf_num: 0,
+        }
+    }
+
+    /// Setupы given `key` on an empty rabbit state, then setupы initialization vector `iv` on it.
+    pub fn new_iv(key: [u8; KEY_BYTE_LEN], iv: [u8; IV_BYTE_LEN]) -> Rabbit {
+        let mut state = Default::default();
+        setup_key(&mut state, key);
+        let master_state = state.clone();
+        setup_iv(&mut state, iv);
+        Rabbit {
+            master_state,
+            state,
+            buf: [0; 16],
+            buf_idx: 0x10,
+            buf_num: 0,
+        }
+    }
+
+    /// Restores master state.
+    pub fn reset(&mut self) {
+        self.state = self.master_state.clone();
+        self.buf_idx = 0x10;
+        self.buf_num = 0;
+    }
+
+    /// Restores master state, than setups initialization vector `iv` on it.
+    pub fn reinit(&mut self, iv: [u8; IV_BYTE_LEN]) {
+        self.state = self.master_state.clone();
+        self.buf_idx = 0x10;
+        self.buf_num = 0;
+        setup_iv(&mut self.state, iv);
+    }
+
+    /// Encrypts bytes of `data` inplace.
+    ///
+    /// Returns:
+    ///
+    /// *   `true` – OK;
+    /// *   `false` – max message length (16 * 2⁶⁴ bytes) was exceeded. `data` is not affected.
+    pub fn encrypt_inplace(&mut self, data: &mut [u8]) -> bool {
+        if !self.check_keystream_len(data.len()) {
+            return false;
+        }
+        for i in 0..data.len() {
+            data[i] ^= self.get_s_byte();
+        }
+        return true;
```

Comment:
One additional clippy failure here:

```suggestion
        true
```